### ISSUE TITLE
No visible indication in the UI that features are not working because of using the wrong Java version

### ISSUE THREAD

**Issue by tjwatson** [eclipse-platform/eclipse.platform#339]
I happened to be using Java 11 as my installation for running Eclipse for some time.  Today I installed m2e from the 2022-03 dev repos and noticed that nothing from m2e was showing up.  Turns out that m2e now requires Java 17, but I got no error popping up that new bundles got installed that require a newer Java version.

I know enough to see if there are unresolved bundles and why they are unresolved but that is not acceptable knowledge for the typical user of Eclipse.  I think the platform should do a quick check for the unresolved bundles to see if any of them require an osgi.ee Java version greater than what is being used to run Eclipse and at least post an error on opening the workspace that states that some features are installed that require a later Java version.

**Comment by iloveeclipse** [comment]
> I installed m2e from the 2022-03 dev repos

How did you do that? I wonder where in the process Eclipse fails to understand that bundles requiring Java 17 can't be used on Java 11. Or was the installation runnimg on 17?

**Comment by laeubi** [comment]
I think @mickaelistria has added support to P2 to warn about that case, but I *think* it does not work if you are using a custom java, so wondering if you have an own java version specified in the `eclipse.ini`?

**Comment by tjwatson** [comment]
> > I installed m2e from the 2022-03 dev repos
> 
> How did you do that? I wonder where in the process Eclipse fails to understand that bundles requiring Java 17 can't be used on Java 11. Or was the installation runnimg on 17?

I always use a recent I-Build build from the Eclipse Project (not an EPP).  I'm on Linux so use update-alternatives/alternatives to switch between the different java versions I have downloaded.  Here my system Java version was set to Java 11 and I used "Install New Software..." and I pointed it to the the site https://download.eclipse.org/releases/2023-03 and selected M2E to install.

Sorry had a typo, I'm using 2023-03 URL to install.

I don't believe this is considered a "custom" java in this case.  At least it should be considered that IMO.

**Comment by mickaelistria** [comment]
@laeubi what you meantion here seems indeed probable: I suspect that @tjwatson has JustJ 17 installed thus p2 thinks everything is fine; but if the Java version is overridden in the eclipse.ini, the checks happening during build aren't much helping.
Aren't there some messages in the log about that?

**Comment by iloveeclipse** [comment]
@tjwatson : which original SDK version are you using?
I've tried to use (almost) latest SDK build with Java 11 and install M2E:

![image](https://user-images.githubusercontent.com/964108/221603350-1c176408-14b7-4ad8-8979-c7e3d0918638.png)

but that fails at very beginning
![image](https://user-images.githubusercontent.com/964108/221603531-273c8d73-2f45-4773-bca6-6cc884d0076e.png)

**Comment by merks** [comment]
I thought @mickaelistria implementation just used the information from the running IDE.  This was implemented a while back when we were planning ahead for an EPP release where Java 17 would be required by the next release:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=483383

**Comment by iloveeclipse** [comment]
@tjwatson: I wasn't able to install at all, and it was immediately visible to me :-)

**Comment by tjwatson** [comment]
Yes, I tried again with a fresh download of I20230226-1800 and I get the expected error from p2.  I've no idea how it didn't fail with my long lived installation that I've been updating from build to build for the past several months.  Perhaps there was some state my long standing install was in that made p2 think Java 17 was there.  Not sure I can go back and find out easily.

**Comment by tjwatson** [comment]
I think I know what happened.  I must have started the eclipse instance while Linux `alternatives` for java was set to Java 17.  But then after that started I probably switched `alternatives` to Java 11.  From there I did the install of m2e while the Eclipse instance was Java 17, so it installed correctly.  Then when I restarted after install it launched with the system Java 11 and the bundles failed to resolve.  This is entirely feasible given that I keep my Eclipse instances running for a long time.  Usually on restart when I do my weekly update.

Nothing wrong with the p2 error here.  This is really about displaying an error when switching Java versions after install.  That is probably not as high priority of an issue as I originally thought.  I still think it may be worth a visible error of some sorts when switching Java versions like that.
